Source/Dafny/Plugin.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Dafny.Plugins;
+
+namespace Microsoft.Dafny;
+
+/// <summary>
+/// This class wraps an assembly and an extracted configuration from this assembly,
+/// The configuration provides the methods to parse command-line arguments and obtain Rewriters 
+/// </summary>
+public record Plugin(Assembly Assembly, string[] Args) {
+  public PluginConfiguration PluginConfiguration { get; init; } = LoadConfiguration(Assembly, Args);
+
+  public static Plugin Load(string pluginPath, string[] args) {
+    return new Plugin(Assembly.LoadFrom(pluginPath), args);
+  }
+
+  class AutomaticPluginConfiguration : PluginConfiguration {
+    private Func<ErrorReporter, Rewriter>[] rewriters;
+    public AutomaticPluginConfiguration(Func<ErrorReporter, Rewriter>[] rewriters) {
+      this.rewriters = rewriters;
+    }
+
+    public override Rewriter[] GetRewriters(ErrorReporter errorReporter) {
+      return rewriters.Select(funcErrorReporterRewriter =>
+        funcErrorReporterRewriter(errorReporter)).ToArray();
+    }
+  }
+
+  public static IEnumerable<System.Type> GetConfigurationsTypes(Assembly assembly) {
+    return assembly.GetTypes()
+      .Where(t => t.IsAssignableTo(typeof(PluginConfiguration)));
+  }
+
+  private static System.Type[] CheckPluginForRewriters(Assembly assembly) {
+    System.Type[] rewriterTpes = assembly.GetTypes().Where(t =>
```

Comment:
For the sake of consistency within the same class (single letter identifiers), I'd use `type` instead of `t` as the identifier name.